src/lib/httpd.ts

```diff
@@ -0,0 +1,184 @@
+import { derived, get, writable } from "svelte/store";
+
+import { HttpdClient } from "@httpd-client";
+import * as mutexExecutor from "@app/lib/mutexExecutor";
+
+export interface Session {
+  id: string;
+  publicKey: string;
+}
+
+export type HttpdState =
+  | { state: "stopped" }
+  | { state: "running" }
+  | {
+      state: "authenticated";
+      session: Session;
+    };
+
+const HTTPD_STATE_STORAGE_KEY = "httpdState";
+
+const store = writable<HttpdState>({ state: "stopped" });
+export const httpdStore = derived(store, s => s);
+
+const api = new HttpdClient({
+  hostname: "127.0.0.1",
+  port: 8080,
+  scheme: "http",
+});
+
+let pollHttpdStateHandle: number | undefined = undefined;
+
+function update(state: HttpdState) {
+  window.localStorage.setItem(HTTPD_STATE_STORAGE_KEY, JSON.stringify(state));
+  store.set(state);
+}
+
+export async function authenticate(params: {
+  id: string;
+  signature: string;
+  publicKey: string;
+}): Promise<boolean> {
+  await disconnect();
+
+  try {
+    await api.session.update(params.id, {
+      sig: params.signature,
+      pk: params.publicKey,
+    });
+    update({
+      state: "authenticated",
+      session: { id: params.id, publicKey: params.publicKey },
+    });
```

Comment:
This can race with the interval function since both touch the same state. In this case we can’t use `mutexExecutor` though. With `mutexExecutor` a call “aborts” any previous calls. But here we want `authenticate()` to finish and block any updates. This requires a proper mutex.

```diff
diff --git a/src/lib/httpd.ts b/src/lib/httpd.ts
index cc09fb8..38f69da 100644
--- a/src/lib/httpd.ts
+++ b/src/lib/httpd.ts
@@ -2,6 +2,7 @@ import { derived, get, writable } from "svelte/store";
 
 import { HttpdClient } from "@httpd-client";
 import * as mutexExecutor from "@app/lib/mutexExecutor";
+import { Mutex } from "async-mutex";
 
 export interface Session {
   id: string;
@@ -34,6 +35,8 @@ function update(state: HttpdState) {
   store.set(state);
 }
 
+const stateMutex = new Mutex();
+
 export async function authenticate(params: {
   id: string;
   signature: string;
@@ -41,26 +44,27 @@ export async function authenticate(params: {
 }): Promise<boolean> {
   await disconnect();
 
-  try {
-    await api.session.update(params.id, {
-      sig: params.signature,
-      pk: params.publicKey,
-    });
-    update({
-      state: "authenticated",
-      session: { id: params.id, publicKey: params.publicKey },
-    });
+  return stateMutex.runExclusive(async () => {
+    try {
+      await api.session.update(params.id, {
+        sig: params.signature,
+        pk: params.publicKey,
+      });
+      update({
+        state: "authenticated",
+        session: { id: params.id, publicKey: params.publicKey },
+      });
 
-    return true;
-  } catch (error) {
-    if (error instanceof Error && error.message === "Failed to fetch") {
-      update({ state: "stopped" });
-    } else {
-      update({ state: "running" });
+      return true;
+    } catch (error) {
+      if (error instanceof Error && error.message === "Failed to fetch") {
+        update({ state: "stopped" });
+      } else {
+        update({ state: "running" });
+      }
+      return false;
     }
-
-    return false;
-  }
+  });
 }
 
 export async function disconnect() {
@@ -122,14 +126,16 @@ function pollSession() {
   }
 
   pollHttpdStateHandle = window.setInterval(async () => {
-    const state = await pollExecutor.run(async () => {
-      const httpd = get(httpdStore);
-      return checkState(httpd);
+    await stateMutex.runExclusive(async () => {
+      const state = await pollExecutor.run(async () => {
+        const httpd = get(httpdStore);
+        return checkState(httpd);
+      });
+      if (state === undefined) {
+        return;
+      }
+      update(state);
     });
-    if (state === undefined) {
-      return;
-    }
-    update(state);
   }, 10_000);
 }
```